File: akit/src/main/java/org/littletonrobotics/junction/networktables/NT4Publisher.java
// Copyright (c) 2021-2025 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

package org.littletonrobotics.junction.networktables;

import edu.wpi.first.networktables.*;
import java.util.HashMap;
import java.util.Map;
import org.littletonrobotics.junction.LogDataReceiver;
import org.littletonrobotics.junction.LogTable;
import org.littletonrobotics.junction.LogTable.LogValue;

/** Publishes log data using NT4. */
public class NT4Publisher implements LogDataReceiver {
  private final NetworkTable akitTable;
  private LogTable lastTable = new LogTable(0);
  private final IntegerPublisher timestampPublisher;
  private final Map<String, GenericPublisher> publishers = new HashMap<>();
  private final Map<String, String> units = new HashMap<>();

  /** Creates a new NT4Publisher. */
  public NT4Publisher() {
    akitTable = NetworkTableInstance.getDefault().getTable("/AdvantageKit");
    timestampPublisher =
        akitTable.getIntegerTopic(timestampKey.substring(1)).publish(PubSubOption.sendAll(true));
  }

  public void putTable(LogTable table) {
    // Send timestamp
    timestampPublisher.set(table.getTimestamp(), table.getTimestamp());

    // Get old and new data
    Map<String, LogValue> newMap = table.getAll(false);
    Map<String, LogValue> oldMap = lastTable.getAll(false);

    // Encode new/changed fields
    for (Map.Entry<String, LogValue> field : newMap.entrySet()) {
      // Check if field has changed
      LogValue newValue = field.getValue();
      if (newValue.equals(oldMap.get(field.getKey()))) {
        continue;
      }

      // Create publisher if necessary
      String key = field.getKey().substring(1);
      String unit = field.getValue().unitStr;
      GenericPublisher publisher = publishers.get(key);
      if (publisher == null) {
        publisher =
            akitTable
                .getTopic(key)
                .genericPublish(field.getValue().getNT4Type(), PubSubOption.sendAll(true));
        publishers.put(key, publisher);

        // Set initial unit
        if (unit != null) {
          akitTable.getTopic(key).setProperty("unit", "\"" + unit + "\"");
          units.put(key, unit);
        }
      }

      // Check if unit changed
      if (unit != null && !unit.equals(units.get(key))) {
        akitTable.getTopic(key).setProperty("unit", "\"" + unit + "\"");
        units.put(key, unit);
      }

      // Write new data
      switch (field.getValue().type) {
        case Raw:
          publisher.setRaw(field.getValue().getRaw(), table.getTimestamp());
          break;
        case Boolean:
          publisher.setBoolean(field.getValue().getBoolean(), table.getTimestamp());
          break;
        case BooleanArray:
          publisher.setBooleanArray(field.getValue().getBooleanArray(), table.getTimestamp());
          break;
        case Integer:
          publisher.setInteger(field.getValue().getInteger(), table.getTimestamp());
          break;
        case IntegerArray:
          publisher.setIntegerArray(field.getValue().getIntegerArray(), table.getTimestamp());
          break;
        case Float:
          publisher.setFloat(field.getValue().getFloat(), table.getTimestamp());
          break;
        case FloatArray:
          publisher.setFloatArray(field.getValue().getFloatArray(), table.getTimestamp());
          break;
        case Double:
          publisher.setDouble(field.getValue().getDouble(), table.getTimestamp());
          break;
        case DoubleArray:
          publisher.setDoubleArray(field.getValue().getDoubleArray(), table.getTimestamp());
          break;
        case String:
          publisher.setString(field.getValue().getString(), table.getTimestamp());
          break;
        case StringArray:
          publisher.setStringArray(field.getValue().getStringArray(), table.getTimestamp());
          break;
      }
    }

    // Update last table
    lastTable = table;
  }
}

File: akit/src/main/native/cpp/networktables/NT4Publisher.cpp
// Copyright (c) 2021-2026 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

#include <networktables/GenericEntry.h>
#include <wpi/json.h>
#include "akit/networktables/NT4Publisher.h"

using namespace akit::nt;

NT4Publisher::NT4Publisher() : akitTable {
		::nt::NetworkTableInstance::GetDefault().GetTable("/AdvantageKit") }, timestampPublisher {
		akitTable->GetIntegerTopic(TIMESTAMP_KEY.substr(1)).Publish( {
				.sendAll = true }) } {
}

void NT4Publisher::PutTable(LogTable &table) {
	timestampPublisher.Set(
			units::microsecond_t { table.GetTimestamp() }.value(),
			units::microsecond_t { table.GetTimestamp() }.value());

	std::unordered_map < std::string, LogTable::LogValue > newMap =
			table.GetAll(false);
	std::unordered_map < std::string, LogTable::LogValue > oldMap =
			lastTable.GetAll(false);

	for (const auto &field : newMap) {
		if (field.second == oldMap.at(field.first))
			continue;

		std::string key = field.first.substr(1);
		std::string unit = field.second.unitStr;
		auto publisher = publishers.find(key);
		if (publisher == publishers.end()) {
			publisher =
					publishers.emplace(key,
							akitTable->GetTopic(key).GenericPublish(
									field.second.GetNT4Type(),
									{ .sendAll = true })).first;
			if (!unit.empty()) {
				akitTable->GetTopic(key).SetProperty("unit",
						"\"" + unit + "\"");
				units[key] = unit;
			}
		}

		if (!unit.empty() && unit != units.at(key)) {
			akitTable->GetTopic(key).SetProperty("unit", "\"" + unit + "\"");
			units[key] = unit;
		}

		switch (field.second.type) {
		case LogTable::LoggableType::Raw: {
			auto value = field.second.GetRaw();
			publisher->second.SetRaw(std::span {
					reinterpret_cast<uint8_t*>(value.data()), value.size() },
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		}
		case LogTable::LoggableType::Boolean:
			publisher->second.SetBoolean(field.second.GetBoolean(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		case LogTable::LoggableType::BooleanArray: {
			auto value = field.second.GetBooleanArray();
			publisher->second.SetBooleanArray(std::vector<int> { value.begin(),
					value.end() },
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		}
		case LogTable::LoggableType::Integer:
			publisher->second.SetInteger(field.second.GetInteger(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		case LogTable::LoggableType::IntegerArray: {
			auto value = field.second.GetIntegerArray();
			publisher->second.SetIntegerArray(std::vector<int64_t> {
					value.begin(), value.end() },
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		}
		case LogTable::LoggableType::Float:
			publisher->second.SetFloat(field.second.GetFloat(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		case LogTable::LoggableType::FloatArray:
			publisher->second.SetFloatArray(field.second.GetFloatArray(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		case LogTable::LoggableType::Double:
			publisher->second.SetDouble(field.second.GetDouble(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		case LogTable::LoggableType::DoubleArray:
			publisher->second.SetDoubleArray(field.second.GetDoubleArray(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		case LogTable::LoggableType::String:
			publisher->second.SetString(field.second.GetString(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		case LogTable::LoggableType::StringArray:
			publisher->second.SetStringArray(field.second.GetStringArray(),
					units::microsecond_t { table.GetTimestamp() }.value());
			break;
		}
	}

	lastTable = table;
}
